app/controllers/whats_app/webhook_controller.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module WhatsApp
+  class WebhookController < ApplicationController
+    skip_before_action :require_login, :verify_authenticity_token
+    UNSUCCESSFUL_DELIVERY = %w[undelivered failed].freeze
+
+    def message
+      adapter = WhatsAppAdapter::Inbound.new
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTRIBUTOR) do |whats_app_phone_number|
+        exception = WhatsAppAdapter::UnknownContributorError.new(whats_app_phone_number: whats_app_phone_number)
+        ErrorNotifier.report(exception)
+      end
+
+      adapter.on(WhatsAppAdapter::RESPONDING_TO_TEMPLATE_MESSAGE) do |contributor|
+        contributor.update!(whats_app_message_template_responded_at: Time.current, whats_app_template_message_sent_at: nil)
+        message = contributor.received_messages.first
```

Comment:
Yeah, I think that when you added this association 3 years ago, it was just assumed that a created message would have been received by the contributor since we have an `after_create_commit` that calls our adapters `send!(message)` method to attempt to send the message out.

In this case, we don't automatically send out the message though, we first check to see if a freeform message is permitted by Twilio/WhatsApp, then either send a template and wait for a response giving us permission to send the message, or if it is permitted, we send the message out.

This results in some interesting semantics where we have a list of "received messages", which might have never been sent out if the contributor is not recently active and/or gives us permission to send the message out.

We also have the case where the contributor does not receive the message because the job to send it out fails for one reason or another. I know of cases where a contributor has never received a message from us because their channel id was invalid and yet you wouldn't know this by looking at the `contributor.received_messages`.

I would be in favor of changing the semantics of this if it also meant having another attribute that would be set when the message is actually successfully delivered to the contributor so we could distinguish between received messages and outbound messages.